src/compose/app.ts

```diff
@@ -200,19 +200,14 @@ class AppImpl implements App {
 					}),
 				);
 			}
-			// Current & target should be the same appId, but one of either current
-			// or target may not have any services, so we need to check both
-			const allServices = this.services.concat(target.services);
-			if (
-				allServices.length > 0 &&
-				allServices.some((s) =>
-					state.locksTaken.isLocked(s.appId, s.serviceName),
-				)
-			) {
+
+			// If the app still has a lock, release it
+			if (state.lock != null) {
 				// Release locks for all services before settling state
 				steps.push(
 					generateStep('releaseLock', {
 						appId: target.appId,
+						lock: state.lock,
 					}),
 				);
 			}
```

Comment:
What happens in a case where the Supervisor dies suddenly when there are no steps expected in the state funnel? If there are no state steps, no `takeLock` steps are ever generated as `appsToLock` is empty for the current appId. So, `state.lock` would be null and no `releaseLock` step would ever be generated. Is this a possible condition on a device?

Reply:
You are right, the `releaseLock` step needs to run every time the state settles and remove any supervisor owned locks even if there is a lock being held. I'll make that change